### ISSUE TITLE
Display Parent Namespace Nodes for Current Namespace in Navi Bar

### ISSUE THREAD

**Issue by wmjordan** [wmjordan/Codist#67]
As @fitdev 's suggestion in issue #64 .

**Comment by wmjordan** [comment]
> I think there is a good way to have multiple search boxes (one for each namespace drop-down) that provide value and avoid confusion. Basically for each namespace dropdown the search would be limited to the appropriate namespace, so that it can search the whole project, but only look inside a specific namespace (for which the dropdown is being displayed). To make this clear, you can change title to: "Search Declaration in [Namespace Goes Here]".
> So, I think you can salvage the current design, avoid confusion, and provide more granular functionality - all at the same time! In fact, it's easy to imagine a scenario where user wants to limit search results for say "con..." to only a particular namespace, as otherwise he/she would be overwhelmed by too many results from everywhere, since "con" is a frequently occurring text.
> Oh, and of course there should be no option to search the current file vs. the whole project for the grayed/dimmed namespaces, since those are not in the active file in the first place. So you would just have the search box w/o file/project buttons on the right.

@fitdev, thank you for the detailed description of the future design.
It is not so difficult but reasonable to have separated search boxes for each level of the namespace hierarchies, limiting the search scope to direct members under the corresponding namespace.
The only concern is the placement of the current *Search Declaration* box and its accompanying filter buttons. I quite enjoy the search box in the recent beta which searches declarations in the current document initially when it is pops up--you can even trigger it with hotkey Ctrl+` in the beta version.
Is it a good idea to place them under the root namespace's drop-down menu? If so, there will be two search boxes in that menu.

**Comment by fitdev** [comment]
I am not sure I understood your question. My thinking was that for a type `CoreNs.Geometry.Shapes.Point2D` for example, the Navi Bar would display:

- { } Core (Dimmed, root namespace), on drop-down: lists all its direct subnamespaces and types; the search box would search within the entire `Core` namespace.
- { } Geometry (Dimmed, parent namespace), on drop-down: lists all its direct subnamespaces and types; the search box would search within the `Core.Geometry` namespace, excluding `Core.Other` namespaces.
- { } Shapes (Normally colored), on drop-down: lists all its direct subnamespaces and types; the search box would search within the `Core.Geometry.Shapes` namespace, excluding `Core.Geometry.Other` namespaces when the _Current Project_ option is toggled; or it would just search within the active document. Note that the 2 items above do not have the option to search project vs. document because the active document does not have any direct "children" of those namespaces.
- Point2D (Type name)
- ...

> I quite enjoy the search box in the recent beta which searches declarations in the current document initially when it is pops up

I agree, it makes sense as the default.

But I did not quite understand why you would need to have 2 search boxes in the root namespace?

**Comment by wmjordan** [comment]
Understood.
I am afraid that the above design can cause confusion, since there can be so many search boxes in those pop-ups. Maybe it is not so easy for a programmer to figure out which one provides the current _Search Declaration_ function.

**Comment by fitdev** [comment]
> Maybe it is not so easy for a programmer to figure out which one provides the current Search Declaration function.

Oh I see! For some reason I though the current design _limited_ search results to a given namespace already, but looks like it searches _all namespaces_ within a project. Sorry, that confused me.

If you want to retain that functionality then I think the best design would be to have a dedicated search box on the Navi Bar itself, perhaps aligned to the right. Although, to be honest I am not sure how useful that functionality )to search all namespaces within the entire project) would be with regards to the Navi Bar, since VS already offers similar thing with "Go To All...".

What I thought would be a true differentiator for the Navui Bar (when compared with vanilla VS experience) is...

For the namespace whose type(s) is/are present in the active file, the Navi Bar item is rendered normally and its search box includes ability to either search document only, or the whole project, _but within that namespace_. And you select which one you want by clicking on the appropriate button on the right side of the search box.

But, once "parent" namespaces are added as additional Navi Bar's item (to the left of the current design's first item), then these new Navi Bar's items would have their own search boxes that would search within their respective namespaces. Here I mean only the actual buttons/items on the Navi Bar itself, not within the dropdowns!

So, basically there would be one search box _per namespace level_ that restricts search results to only come from that namespace. I don't think that's confusing or a lot of search boxes, as typically there are no more than 3-4 levels of Namespace nesting, and usually just 2.

**Comment by wmjordan** [comment]
Good to read your response.

The _search project_ function in the _Navi Bar_ does almost the same thing as "_Go to All_" command. It is possible to remove it.
However, the _search document_ function, I think, is more convenient for searching within active document than using the "Go to..." panel offered by VS. @fantasyz provided his suggestions in #56 and he appeared to use that function quite frequently.

**Comment by fantasyz** [comment]
I myself enjoy and use the local file search very lot. Please don't remove it. I think it is better to be able to trigger different search scope (local, project) with different shortcuts. People who use search symbol on whatever scope, just use different shortcut to trigger the corresponding search box. Optionally, it would be great if I can use a key, say TAB key,  to circular toggle between them.

Another thing I would be happy to see is the ability to search member of a class ACROSS multiple files. e.g. partial class on multiple files. The search local file function in this case are not able to search member of same (partial ) class on another file but the search project are returning too much result. Class member search is something in between.

The search project maybe redundant to the native "Go To All..." but I think it can be improved to replace it, esp the Go To All is not presenting the result in a very good way. (too much noise). No to mention it is not easy to have improvement on these not so important things from Micro$oft.

**Comment by fitdev** [comment]
> However, the search document function, I think, is more convenient for searching within active document than using the "Go to..." panel offered by VS. @fantasyz provided his suggestions in #56 and he appeared to use that function quite frequently.

Totally agree! I was talking _only_ about the "Search Project" function.

**Comment by fitdev** [comment]
> Another thing I would be happy to see is the ability to search member of a class ACROSS multiple files. e.g. partial class on multiple files.

I agree! That would be very useful. So, perhaps you could just replace the current whole-project option with the partial types option. And those who do really need search the entire project could use VS's "Go To All".

This, together with the ability to search specific namespaces, as I suggested above, should cover all possible use cases!

**Comment by wmjordan** [comment]
@fantasyz Thank you for join in this discussion.

> I myself enjoy and use the local file search a lot. Please don't remove it.

Me too. So it will be there.
The problem is where to place it if parent namespace nodes are also placed on the Navi Bar. If my understanding was right, @fitdev suggested the file search box placed at the right side of the Navi Bar.

> I would be happy to see is the ability to search member of a class ACROSS multiple files. e.g. partial class on multiple files.

You don't have to use _would_. The publicly released Codist had already provided the capability to search across partial types on multiple files.
![default](https://user-images.githubusercontent.com/12182730/50342411-eeeb3480-055d-11e9-9235-d4d9c234719f.PNG)
Input something in the text box on the drop down menu of a type, the search will narrow down to that type. The above screenshot were taken for the `CodeAnalysisHelper` class in source code files of _Codist_. That class were a `partial` class spanning across two files. The gray items were in external code file and the black one was in the current code file.

@fitdev 
> perhaps you could just replace the current whole-project option with the current namespace option (which would include partial types as well)

Yes, I am considering this.

**Comment by fitdev** [comment]
> For instance, it is possible to add member type filters to help narrow down the search results.

Absolutely, that would be great! Limiting results by both namespace and/or member/type kind is a nice refinement! However, as for the project scope search (without namespace filtering that is), I still do not see how that would be different from the "Go To All..." - that feature lets you filter member as well (fields, properties, methods, etc.).

As I would imagine it:

- The current-file based search would appear as a separate searchbox (say on the right). It would probably be the most frequently used option). Possibly, it should, by default include all the members from partial types defined in the current file as well.
- "Parent" namespace, current namespace, and current type Navi Bar items _would each have their own_ search box, with results limited to that specific namespace/type, but _searching across the entire project_ (not just the active file)
- Finally for a true, unrestricted project-wide search, VS's "Go To All" feature would be used. It allows to filter by member/type kind as well.

Doesn't it cover all the possibilities?

**Comment by fantasyz** [comment]
> You don't have to use _would_. The publicly released Codist had already provided the capability to search across partial types on multiple files.

Oh, I didn't realize that. I think I use local search too much and forgot about the existence of this class search. Is there a shortcut for that search box? or do I always need to use mouse click to trigger it? 
Btw, it is possible to have the drop down list longer to show more items? I prefer it to be as long as possible when needed.

About the position of Search Symbol box, it can also be place at the center of the screen as a pop up dialog like Visual Assist. A totally stand alone component. Not necessarily belong to Navi Bar. This also give you more UI space to add lots of future functions.

![alt text](https://www.wholetomato.com/images/learn/top10/findSymbol.png "VAX")

**Comment by fitdev** [comment]
> I prefer it to be as long as possible when needed.

Yeah, that would be nice, I agree.

> About the position of Search Symbol box, it can also be place at the center of the screen as a pop up dialog like Visual Assist. A totally stand alone component. Not necessarily belong to Navi Bar. This also give you more UI space to add lots of future functions.

Alternatively, you may consider still placing it on the Navi Bar, but as a button by default, which, when clicked / activated, would expand into full searchbox. That way, it will remain easily discoverable, but won't consume valuable space. Also, if VS allows it, you could shrink the VS's "Project Selector" dropdown on the left to give more space to the NaviBar. Right now it is using 33% of the width, when typically it only needs around 10% (depending on the length of the project name).

**Comment by fantasyz** [comment]
>  ...as a button by default, which, when clicked / activated, would expand into full search box.
this is exactly what I think. 

> could shrink the VS's "Project Selector" dropdown on the left to give more space to the NaviBar. Right now it is using 33% of the width, when typically it only needs around 10% (depending on the length of the project name).

I actually don't understand why is it implemented as a dropdown. If navibar is showing code path of current code file, isn't the project always the same for that particular file?  Most of the time, I found it only has one item in the dropdown.

**Comment by fitdev** [comment]
> If navibar is showing code path of current code file, isn't the project always the same for that particular file

No, it's not, at least in 2 cases:

- It could be part of a Shared Project
- It could be a linked file

> Most of the time, I found it only has one item in the dropdown.

That really depends on your solution, and the number of shared vs "real" projects it has.

**Comment by fantasyz** [comment]
Oh I see. So the dropdown is actually showing what projects are sharing the current code file.
Thanks

**Comment by wmjordan** [comment]
@fantasyz 
The _Find Symbol_ windows has been planned for quite some time, weeks before your coming to here. You can find that there is a file named _SymbolFinderWindow_ among the source code files track back its history here.
Nevertheless, it will take me quite some considerable time to implement it and maybe postponed to version 5.

> I prefer it to be as long as possible when needed.
Initially, without the search box and the member filters control, the menu was shown that way. However, there is an issue in the implementation of drop down menu in WPF which fails to calculate the correct possible height after adding those controls.

**Comment by wmjordan** [comment]
@fantasyz 
> I think I use local search too much and forgot about the existence of this class search. Is there a shortcut for that search box?

It is possible to have a shortcut for class search box. I am planning to add some combo shortcuts.
The shortcuts are preceded with Ctrl+\`, then followed by another Ctrl+shortcut key.

1. Ctrl+\`, Ctrl+\`: search active document
2. Ctrl+\`, Ctrl+1: search enclosing class
3. Ctrl+\`, Ctrl+2: search enclosing namespace

**Comment by wmjordan** [comment]
@fitdev 
I noticed that the recent version had adapted the idea of Codist to hack the navigation bar.
Would you like to install it (I suggest you disable Codist since they may conflict) and share your thoughts?

https://github.com/sboulema/CodeNav/issues/92

**Comment by fitdev** [comment]
Tried it, but it crashes every time after I set it to Top, so cannot see how they have implemented it. I did disable Codist but it did not help. Now I am back to Codist...

**Comment by wmjordan** [comment]
Now I am back to the development to this feature request.

But, before starting, I'd love to localize Codist to Chinese before working on this.

I have an idea of the design of the Navi Bar.
The ancestor namespaces will be listed, and there will be a file button at the left most part (see below).

[File] [RootNamespace] ...(more buttons if more namespaces) [Namespace] [Type] [Member]

Clicking the [File] button will pop up the search menu, similar to the one displayed nowadays we click on the namespace button. Clicking on the namespace buttons will bring out a list displaying namespaces and types under it (filterable).

**Comment by fitdev** [comment]
Sounds great! I like it!

**Comment by wmjordan** [comment]
Hi @fitdev ,
Today is the traditional New Year in China. I am on holiday now and I just managed to squeeze some time on this feature.
Please take a look at the new beta.

**Comment by wmjordan** [comment]
@fantasyz Please help test this new beta too ^_^

**Comment by fantasyz** [comment]
Just tried. Its seems that some bars are not working correctly. Only the one in the left most split view works.
![codist](https://user-images.githubusercontent.com/3223242/107774099-87198880-6d79-11eb-952c-8b632018d3e4.png)

**Comment by fantasyz** [comment]
![codist2](https://user-images.githubusercontent.com/3223242/107774677-41a98b00-6d7a-11eb-8556-dab38af42186.png)

Found 2nd issue.
I use Declaration Search on project. A partial class across different files bring out a file selection window. That 2nd popup doesn't close automatically after selection.

**Comment by fantasyz** [comment]
![codist 2](https://user-images.githubusercontent.com/3223242/107774909-977e3300-6d7a-11eb-9a54-e1e6f79820de.png)

Can the search result be sorted in DeclarationSearch when using project scope? Will it be too slow? It For local search, using the default order as in the code file is just fine.

Still hoping to have direct shortcuts to open DeclarationSearch to both Local and Project scope. (Don't want to touch the mouse T_T)

Thank you and Happy New Year

**Comment by fitdev** [comment]
Thank you for having a go at this feature and a Happy Chinese New Year!

It looks really nice, I love the ability to be able to click on the namespace hierarchy in the Navi Bar and see the nested namespaces and types they contain. That's a huge improvement for me and makes NaviBar much more productive!

Also, I am using latest VS Preview 16.9 Preview 4, and I have _not_ experienced the first issues @fantasyz reported. In my case, even in vertical-split layout, both sides show proper NaviBar.

**Comment by fitdev** [comment]
One suggestion though: in addition to (or perhaps instead of (though I like both)) displaying the source file names where a particular type/namespace is defined, it would help to also display project name(s) / assemblies. I mean the information that appears on the tooltip when you hover over various dropdown items on the NaviBar.

It is quite common for namespaces to be defined across _several_ projects / assemblies, so displaying this info is quite useful.

As for displaying the list of files for _namespaces_ (types are usually ok), that list may be long indeed (i.e. easily more than 100 files for some namespaces), so I am not entirely sure if it is useful to display all the source file names where a namespace is defined in a tooltip. Perhaps a better approach would be to add another item to the right-click context menu with the option to list all the source files? For example, in my case the generated tooltips for some of the namespaces are like 30+ lines long with more than a few hundred file names being listed.

**Comment by wmjordan** [comment]
@fantasyz 
> Its seems that some bars are not working correctly. Only the one in the left most split view works.

It is weird. I have never met with this issue before and after the new implementation. The Navi Bar only displays things like the right window when we click outside of the `namespace` scope.
Could you describe how this happened?

> That 2nd popup doesn't close automatically after selection.

I will take a look at it. I seems to me that it is an intentional design so that we can go to another file if we need to do so. We can click away the menu to close it.

> Can the search result be sorted in DeclarationSearch when using project scope?

Actually the result is "sorted" by **the length of the matched names**. So if you type "Match", a member named "Match" will always be placed before "MatchItems", "ExtraMatch", "MatchInCollection", etc.

> direct shortcuts to open DeclarationSearch to both Local and Project scope

At the day when this feature is developed, we can use the shortcut by hitting Ctrl+\` twice.
However, in the recent version of VS 2019, Ctrl+\` has been used to bring out a Power Shell tool window. Thus you have to manually reassign another shortcut to bring out that menu.


@fitdev 
> That's a huge improvement for me and makes NaviBar much more productive!

Thank you for your feedback. I am still trying it out and see how it helps me :)

> As for displaying the list of files for _namespaces_ (types are usually ok), that list may be long indeed (i.e. easily more than 100 files for some namespaces)

Yes, you are right. I took not very long after the test and saw the same problem. I will make changes to it.

**Comment by wmjordan** [comment]
@fantasyz 
I changed the behavior of symbol locations pop up menu to be automatically closed if it was not pinned. Please try it.
Personally I would like that menu to be staying there by default, since a symbol could be in multiple files, if I jumped to the wrong file, I could still go back and try the others without repeating the procedure to bring out that menu.

@fitdev 
I removed the long list of files for `namespace`s and displayed the number of locations and listed containing assemblies instead--this change looked much better for me too. I am sorry that I could not find a way to get back which project a `namespace` may belong to. The best thing I could get was the containing assemblies.

Please try the new beta 6127.

**Comment by fantasyz** [comment]
Tested the update. Here is a little feedback.

> @fantasyz
> 
> > Its seems that some bars are not working correctly. Only the one in the left most split view works.
> 
> It is weird. I have never met with this issue before and after the new implementation. The Navi Bar only displays things like the right window when we click outside of the `namespace` scope.
> Could you describe how this happened?
I actually didn't do anything. It is what happens when I updated Codist and restart VS. Btw, it no longer happens after a 2nd restart.
 
> > That 2nd popup doesn't close automatically after selection.
> 
> I will take a look at it. I seems to me that it is an intentional design so that we can go to another file if we need to do so. We can click away the menu to close it. 
> Personally I would like that menu to be staying there by default, since a symbol could be in multiple files, if I jumped to the wrong file, I could still go back and try the others without repeating the procedure to bring out that menu.


![devenv_egUibaSYiI](https://user-images.githubusercontent.com/3223242/107843901-f8088100-6e09-11eb-8759-31ea2435b12c.png)

I see. I like the original pop up position right next to the drop down more than at the cursor because I don't use the mouse but only the keyboard during the whole process. Now the pop up location is like random to me. It just feel weird when it shop up on another split-view cell or very far away. If want take care of both Keyboard Only and Keyboard/Mouse users, I would suggest let it be a setting to change how it show. e.g. Fixed vs Follow Cursor. 


![search symbol](https://user-images.githubusercontent.com/3223242/107844358-9f3ae780-6e0d-11eb-8404-966be2f2e4fb.png)
Not really a big issue. I would personally prefer class definition get higher a priority than variable declaration?


> > Can the search result be sorted in DeclarationSearch when using project scope?
> 
> Actually the result is "sorted" by **the length of the matched names**. So if you type "Match", a member named "Match" will always be placed before "MatchItems", "ExtraMatch", "MatchInCollection", etc.

I see. 

> 
> > direct shortcuts to open DeclarationSearch to both Local and Project scope
> 
> At the day when this feature is developed, we can use the shortcut by hitting Ctrl+` twice.
> However, in the recent version of VS 2019, Ctrl+` has been used to bring out a Power Shell tool window. Thus you have to manually reassign another shortcut to bring out that menu.

I mean in the current build, I need to press the shortcut and use the mouse to click the buttons to toggle between Local and Project scope. I would like to eliminate the "mouse click". For example, this is what I would like to setup:
[Ctrl+M] to directly open the search declaration with **local** scope. 
[Ctrl+Shift+M] to directly open the search declaration with **project** scope.

**Comment by fitdev** [comment]
> I removed the long list of files for namespaces and displayed the number of locations and listed containing assemblies instead--this change looked much better for me too. I am sorry that I could not find a way to get back which project a namespace may belong to. The best thing I could get was the containing assemblies.

Thank you! It looks much better! A few more suggestions:

1. Perhaps you can include Assembly/Project name/path (whichever is easier for you) for _types_ as well (in addition to including that info for namespaces)?
2. The Files popup that appears when you click on a namepsace in the dropdown is great! Maybe you can include the Assembly/Project name for the source file tooltips as well?
3. And in conjunction with #2, a way to _filter source files based on Project/Assembly_ would be even better (similar to how you have filters for the kinds of types for example). I realize it may require some work as you would need to include a combobox with a list of applicable projects/assemblies.

**Comment by wmjordan** [comment]
> I like the original pop up position right next to the drop down more than at the cursor because I don't use the mouse but only the keyboard during the whole process. Now the pop up location is like random to me.

That's weird. I have not changed the logic about the location where the popup emerges. It should always pop up under the mouse cursor.
In the current implementation, the pop up menu has no context information about how it is bring out, by mouse or by keyboard, from which control. So it is hard to get what you need.

However, during my tests after your reporting this issue, I found that the pop up could emerge on another document group window if we have more than one document groups.

I will take a further look into it.

> I need to press the shortcut and use the mouse to click the buttons to toggle between Local and Project scope

Understood.

> I would personally prefer class definition get higher a priority than variable declaration?

Hmm, it is reasonable. I will try that.

**Comment by wmjordan** [comment]
@fitdev 
Thank you for your...
> A few more suggestions

However, only the first one of them is very possible at this moment.

I will try and see whether there is a way to implement the next two...

**Comment by fitdev** [comment]
> The project/assembly filter really requires some work. Maybe we can specify the project or assembly by using specific notations.
> For instance, "Manage /Console" will search "Manage" in file names and "Console" in project names.

I did not mean implementing the whole filter, although that would be nice. But just the ability to include assembly or project name in the tooltip for the source files as well as types.

**Comment by fitdev** [comment]
Another issue I noticed is that at present the namespace dropdowns show only members (namespaces and types) that re defined _in the same assembly / project_ as the currently opened active document. It would be great if they would show _all user code_ (including types and namespaces defined in _other projects_ that have been opened as part of a solution). Perhaps this can be an option, in the filter toolbar or a global option.

**Comment by fantasyz** [comment]
> 
> 
> > I like the original pop up position right next to the drop down more than at the cursor because I don't use the mouse but only the keyboard during the whole process. Now the pop up location is like random to me.
> 
> That's weird. I have not changed the logic about the location where the popup emerges. It should always pop up under the mouse cursor.
> In the current implementation, the pop up menu has no context information about how it is brought out, by mouse or by keyboard, from which control. So it is hard to get what you need.

Haha. In this case I probably just by luck getting the popup right next to the selection when testing. It feels more comfortable to see the pop up next to the selection like we normally see in other apps. 

![chrome_uLAHOvHSs7](https://user-images.githubusercontent.com/3223242/107851499-720a2b80-6e45-11eb-9747-5170e6104870.png)

For a less complicated fixed position implementation, I would suggest to have it popup just at the center of the visual studio window or remember the last position where I dragged and closed it so that it will always appear on a predictable location. Of course this suggestion is for keyboard intense users like me. Mouse users may still want to popup to appear around the cursor.

**Comment by wmjordan** [comment]
@fitdev 
> the namespace dropdowns show only members (namespaces and types) that re defined _in the same assembly / project_ as the currently opened active document. It would be great if they would show _all user code_ (including types and namespaces defined in _other projects_ that have been opened as part of a solution)

Thank you for your notification.
If it is true, it is not an intentional implementation.
I did not notice this since I did not have projects implementing a same namespace.
I will test this later.

@fantasyz 
> It feels more comfortable to see the pop up next to the selection like we normally see in other apps.

I know that.
As @fitdev has mentioned in the above comments, he suggested that the namespace items in the drop down can be expanded to display defined child namespaces and types.
I am examining the code and see how to implement this.

**Comment by wmjordan** [comment]
@fantasyz 
> I like the original pop up position right next to the drop down more than at the cursor because I don't use the mouse but only the keyboard during the whole process. Now the pop up location is like random to me.

I managed to reproduce your problem. Fixed in the new beta.
If the mouse is not on the clicked menu item, then the popup will show up next to the menu item.

**Comment by wmjordan** [comment]
@fitdev 
I uploaded a new beta version.
In this new version, if a namespace menu in the NaviBar contains sub-namespaces, when we click one of those sub-namespaces, a menu containing types in that sub-namespace will be displayed.
![image](https://user-images.githubusercontent.com/12182730/108057523-b851cb80-708d-11eb-8bf1-8a39ed77c667.png)

In previous builds, files where the sub-namespace is defined are listed. I think it is more helpful to list types instead of files in this scenario. To list files like previous versions, we can right click the sub-namespace item and select the _Go to Definition_ command.

**Comment by fantasyz** [comment]
I found that the dropdown doesn't automatically close on the original tab after I select a symbol that would open another file.

**Comment by fitdev** [comment]
> I think it is more helpful to list types instead of files in this scenario. To list files like previous versions, we can right click the sub-namespace item and select the Go to Definition command.

That makes sense. However, as I suggested in another issue, perhaps at some point in the future this functionality will be implemented using an in-place Tree View instead.

**Comment by wmjordan** [comment]
> I found that the dropdown doesn't automatically close on the original tab after I select a symbol that would open another file.

Yes. I know that.
I am testing with this function and see whether it is more appropriated to keep the menu there since it may serve the purpose of being a symbol browser.

**Comment by fantasyz** [comment]
> 
> 
> > I found that the dropdown doesn't automatically close on the original tab after I select a symbol that would open another file.
> 
> Yes. I know that.
> I am testing with this function and see whether it is more appropriated to keep the menu there since it may serve the purpose of being a symbol browser.

I see. Please add a setting to turn this "feature" off if you manage to keep this design. It is kind of annoying to me.

**Comment by fitdev** [comment]
> I am testing with this function and see whether it is more appropriated to keep the menu there since it may serve the purpose of being a symbol browser.

I think I prefer it to stay open. That way you know what you just clicked on. And you can always auto-close it by clicking on the text area to start typing. But I suppose it should be left as an option, since people have different preferences.

**Comment by fitdev** [comment]
On a related topic, a quick suggestion...

Right now when you click on a namespace in the dropdown, it opens a popup with a list of subnamespaces and types. That's great! But then, if you click once more on a _snamespace_ item _in that popup_, the original popup is closed and a new popup is shown. But sometimes it is useful to leave the original popup open still even when you are openning additional popup by clicking on the child namespaces. So, perhaps if you do `Ctrl+Click` instead of regular click it will _not_ hide the previous popup, i.e. so that it behaves more like a menu.

**Comment by wmjordan** [comment]
> perhaps if you do `Ctrl+Click` instead of regular click it will _not_ hide the previous popup, i.e. so that it behaves more like a menu

Hmm, this makes sense. I will try this way.

**Comment by wmjordan** [comment]
Darn, this feature took only a day to program but a whole week's holiday to fix bugs!

@fantasyz 
> I see. Please add a setting to turn this "feature" off if you manage to keep this design. It is kind of annoying to me.

Considering yours and @fitdev 's opinion, I changed the behavior to close the menu by default. The menu will stay only when the symbol is defined in multiple locations.
I also discovered and fixed a bug which can crash VS when testing your case.

Please download the new beta.

**Comment by fitdev** [comment]
> Darn, this feature took only a day to program but a whole week's holiday to fix bugs!

I hope it is worth the effort in the end, especially for those with large codebases comprising 10s of medium/large projects. Even as it is, I already use it regularly to quickly check things or jump around code.

**Comment by fantasyz** [comment]
> 
> 
> Darn, this feature took only a day to program but a whole week's holiday to fix bugs!
> 
> @fantasyz
> 
> > I see. Please add a setting to turn this "feature" off if you manage to keep this design. It is kind of annoying to me.
> 
> Considering yours and @fitdev 's opinion, I changed the behavior to close the menu by default. The menu will stay only when the symbol is defined in multiple locations and then the location popup will show up.
> I also discovered and fixed a bug which can crash VS when testing your case.
> 
> Please download the new beta.

I think my preference is a little different. The fact is, I don't want the dropdown to stay open at all under all circumstances even if a symbol exist across multiple files. It just doesn't feel right to me that I selected the file I want and the UI still staying there assuming I choose the wrong file waiting for me to re-select. It is even more weird when I switch back to an old tab and suddenly see the already opened dropdown with previous search state on it not to mention it is blocking part of the code. If guess it is affected by how well I know about the project. If I choose the wrong file (which doesn't happen that much), I would rather press a shortcut to bring up the previous search result again and to try the next file. 
By the way, I found that the dropdown of each source file (or tag ) has its own state! which IMO is overdoing and is doing more harm. I generally only care about the last search result when coding. I guess most programmers also just focus on one thing and the 2nd last or even 3rd last search result staying on another tab seems not really useful. If it make sense, I prefer the tabs to share the same search record so that I could always refine my previous search on any of the tab instead of just the last tab. 

Regarding the Pros and Cons of keeping the UI open or having it auto close, if the UI keep it open by default, it save a key press to re-open it in case a wrong file is selected but would require an extra  keypress/click to close it. On the other hand, if the UI auto close, it save a keypress/click to close it but would require an extra keypress/click to reopen it. In summary, which one is better really depends on how well the user know the project/file. The more accurate I could choose my file, the more I would like it to auto-close to skip the extra key press. There is no standard answer which one is better.

I think fitdev's suggestion of using [Ctrl+Cick] (or [Ctrl+Enter] for keyboard user) to flip the default auto close behaviour is great and enough for both worlds.  To take care of all users who want different default and flexibility, provide an option to toggle the default auto-close behaviour and support the Ctrl modifier key to flip that default behaviour in any time. Everyone happy~

**Comment by wmjordan** [comment]
@fantasyz Please try the new beta and see how it goes.

> I found that the dropdown of each source file (or tab ) has its own state! which IMO is overdoing and is doing more harm.

I don't think it a very big problem. Each drop down belongs to their own document window and they shall not share the same state. I have had difficulty getting them in sync before, but when I found a way out, I **deliberately** leave them as it is to now. If it is the problem for you that only one search result is useful, maybe Alt+\\ is the way to go for you.

**Comment by fantasyz** [comment]
> 
> 
> @fantasyz Please try the new beta and see how it goes. The menus now automatically closes until Ctrl is pressed.
> 
> > I found that the dropdown of each source file (or tab ) has its own state! which IMO is overdoing and is doing more harm.
> 
> I don't think it a very big problem. Each drop down belongs to their own document window and they shall not share the same state. I have had difficulty getting them in sync before, but when I found a way out, I **deliberately** leave them as it is to now. If it is the problem for you that only one search result is useful, maybe Alt+\ is the way to go for you.

Ah, I think I am not explaining enough about the share state situation. I agree with you that it is not a big issue BUT only when using the drop down for local members. It is a different story for project-wise search. It is better to have a single common search record for project search but on the other hand it doesn't make sense to have shared the search result for local member search because members are exclusive to the own file. The current same looking UI make us easily to think that they are similar but the best way to handle them are a little different if you understand what I mean.

**Comment by wmjordan** [comment]
> It is a different story for project-wise search. It is better to have a single common search record for project search

Well, I agree with you.
While I was doing project-wide search, I felt that I could brought the search result to the newly opened file after I clicked a search result. What do you think the user interface shall behave in your opinion?

There is a problem about behavioral consistency since the default search scope in the member search function is local member search (within current code file).
If a user switches to another code file and opens the search function on the NaviBar...what shall happen next?

**Comment by fantasyz** [comment]
@wmjordan 
Wow, I don't even remember I suggested that long ago. About the UI suggestion, it is just another way to present the data and is not really that different functionally at the moment unless more related features are added in the future. I think the main focus here is still to have the project-search share the same history so that the last search can be continue from another tab. 

Simply put
For Navi drop down with local member search >>> use the current implementation for the search history
For Navi drop down with project symbol search >>> use a shared search history for the project

No UI change needed atm. I think that's great enough for the current build.

**Comment by wmjordan** [comment]
Hi @fantasyz ,

On keyboard programmers' side, your suggestion is Okay.
But if we operate with a mouse, each time when we click the search menu button, it pops up the local member search.

I will see whether it is possible to auto-fill the search box if the search menu is triggered by keyboard shortcut or the menu command under the Edit menu.

> The fact is, I don't want the dropdown to stay open at all under all circumstances even if a symbol exist across multiple files

I am sorry that I has been messing with the wrong piece of code and your situation had never gotten improved--I though I had. I will address this in the upcoming beta.

**Comment by fantasyz** [comment]
@wmjordan  
> But if we operate with a mouse, each time when we click the search menu button, it pops up the local member search.
I think the ideal case is to still show the local member search history since it member search is default for mouse triggering.
When user toggle the UI tothe project scope, just swap the search field with project search history if there is any.

> > The fact is, I don't want the dropdown to stay open at all under all circumstances even if a symbol exist across multiple files
> 
> I am sorry that I have been messing with the wrong piece of code and your situation had never gotten improved--I though I had. I will address this in the upcoming beta.
Yup, I just updated to build 6210 and is confirmed everything is still staying open when selected an option jumping to another file. Looking forward for the next update and thank you for improving this aspect of the Navi Bar. It helps a lot.

**Comment by wmjordan** [comment]
Friends, please try the new beta which has fixed some terrible crashes and has improved working experience for keyboard programmers.

Darn, this feature has hogged my whole holiday and a weekend...

**Comment by fantasyz** [comment]
@wmjordan Nice, I like it. 
A little feedback! Not a bit issue but when I re-open a project search drop-down on another tab. There is a chance the cursor is at the beginning of the text instead of selecting the whole text. 

![devenv_GdgHlMdiof](https://user-images.githubusercontent.com/3223242/108625433-e5cfb800-7485-11eb-8a94-b887440ce8de.png)

**Comment by fantasyz** [comment]
Found more issues

1. The 2nd popup is not auto-closing with the main drop down when pressed the ESC.
2. When the 2nd popup appear, it better to default the selection to the 1st item instead of nothing selected.

![Not closeing when pressing Esc](https://user-images.githubusercontent.com/3223242/108634469-a9678080-74b4-11eb-9957-bf3fcc3fc895.gif)

3. When reopen a search, although the textbox is filled with previous text, the previous selection is no longer selected.

![Default to 1st option 1](https://user-images.githubusercontent.com/3223242/108634516-cf8d2080-74b4-11eb-96df-80134b6c094f.png)

**Comment by wmjordan** [comment]
> There is a chance the cursor is at the beginning of the text instead of selecting the whole text.

Hmm, I also noticed this. Will be done in the next beta.


> 1. not auto-closing with the main drop down when pressed the ESC

Oh, this is annoying. I will see how to do with it.

> 2. When the 2nd popup appear, it better to default the selection to the 1st item instead of nothing selected

Don't care about this. You can always press Enter key to activate the first item in the list if the caret is in the text box.

> 3. When reopen a search, although the textbox is filled with previous text, the previous selection is no longer selected

If you activate an item in the search result, then in the newly created document window, it does not help you any more to select the same item which is already in the current document.

**Comment by fantasyz** [comment]
> > 2. When the 2nd popup appear, it better to default the selection to the 1st item instead of nothing selected
> Don't care about this. You can always press Enter key to activate the first item in the list if the caret is in the text box.

True, for the 1st item. However, for the 2nd item, I need to press down twice before pressing enter.

> >  3. When reopen a search, although the textbox is filled with previous text, the previous selection is no longer selected
> 
> If you activate an item in the search result, then in the newly created document window, it does not help you any more to select the same item which is already in the current document.

However, it does help continue to the next/prev one with fewer "Up/Down" key press. This little advantage may go unnoticed if you mainly use the mouse to select the items. 

No big deal. These little refinements are for perfection. 
Just a little more key press. I can live with it.

**Comment by wmjordan** [comment]
@fantasyz There's a new beta now.
The "popup auto-closing with the main drop down when pressed the ESC" is not implemented. However, you can press ESC some more times to close the remained ones.

@fitdev 
I am dealing with the hell mode. Lots of crashes and slow-down.
Some more pieces of code have to be rewritten.

**Comment by wmjordan** [comment]
In the current implementation, if projects define the same namespace but they do not share a same code file, there will be duplicated namespace items in the drop-down.
<img width="783" alt="shot" src="https://user-images.githubusercontent.com/12182730/108928985-272ab800-767e-11eb-96a2-5a71d599ae27.png">

**Comment by fitdev** [comment]
It's interesting that I have experienced just the one crash I reported. But I have not used it extensively yet. I will keep an eye for crashes and report them when I encounter any.

I did not think that it would require so much effort. If you feel overwhelmed, perhaps for the first release you can concentrate your efforts on showing only those items that are referencable from a given projects (as opposed to showing all items available in a solution, which you can implement later, perhaps once you have thought of a better strategy to deal with this larger number of items).

**Comment by wmjordan** [comment]
> showing only those items that are referencable from a given projects

Maybe there will be two statuses for the menu: By default, only the referenced symbols are shown, and at the bottom of the menu there's a button to search for unreferenced symbols.

But crashes can still occur when Roslyn is heavily burdened with large projects.
Some refactoring has to be done to eliminate them.

**Comment by fitdev** [comment]
> Maybe there will be two statuses for the menu: By default, only the referenced symbols are shown, and at the bottom of the menu there's a button to search for unreferenced symbols.

That would be great! I was just saying that you could prioritize your work on the referencable items, if the other case is too complicated at the moment.

> Some refactoring has to be done to eliminate them.

I see. Hopefully this is worth it in the end, as it can provide a more solid foundation upon which to extend the symbol browsing capabilities even more.

**Comment by wmjordan** [comment]
Currently I found that Roslyn was not behaving as what we thought it should be.
It somehow created a temporary "assembly" with a random name while I was inspecting a namespace symbol, but actually the symbol was in the current project and consequently this led Codist thought that the symbol was in an external, unreferenced project.

**Comment by wmjordan** [comment]
@fitdev 
I found a way to eliminate the duplicated items in the namespace drop-down on the NaviBar.
Please test the new beta and see whether the display is right.

**Comment by fitdev** [comment]
Thank you for the new version! Seems to be working well. I have not seen any duplicate entries. A few minor suggestions:

1. Is it possible to introduce some way to identify (highlight) types that are defined in the current file / project within a given namespace as opposed to the types defined in other files / projects (within identically named namespace). I have some files that contain multiple types within the same namespace, and it would be nice to see which of those types are defined in the current file without scrolling the code - just by looking at the NaviBar's namespace dropdown.

In fact maybe this can be generalized to the following filter modes:
- Current file only
- Current project
- Current project + all projects it depends on
- Entire solution

Perhaps with an option to remember these settings for a given namespace?
This will also address the performance concerns, as the default could be current project only for example.

2. I seem to get inconsistent results with the number of items being listed for namespace dropdowns. Eventually I seem to get all members (after having dropped down the Namespace item several times from different files and projects). I understand that it takes time for this info to be gathered and computed. So the suggestion would be to show some kind of indication - an icon, label, progress bar that informs the user that the items are _still being gathered_. This is similar to how VS Intellisense does it when it is trying to display large number of Intellisense items.

**Comment by wmjordan** [comment]
> I seem to get inconsistent results with the number of items being listed for namespace dropdowns. ... So the suggestion would be to show some kind of indication - an icon, label, progress bar that informs the user that the items are _still being gathered_.

This could happen.
I also hope to have some clue that Roslyn is busy. However, I could not find any event or property to poll its status.

**Comment by wmjordan** [comment]
Hey friends, I am going to close this request and release the new version, which is imperfect, but should be better than existing one in the VS market place.

Have you ever met with any new crash?

> A few minor suggestions

They will be addressed in the future version.

**Comment by fantasyz** [comment]
I do experience VS crash but the stacktrace doesn't have any trace related to Codist. I guess it probably is irrelevant.

**Comment by fantasyz** [comment]
One of the crash I can reproduce almost 100% is that when I try to drag a tab from one side to another side of the split view cell. However, it does not happen if I use shortcut to move the tab. Do you also have this issue in your VS?

**Comment by wmjordan** [comment]
Thank you for replying.
I don't have that issue on my VS.

I work with my laptop which has a not-so-big screen thus it is too luxurious for me to have split view cell.
However, I tried quite a few drags of tabs among various document groups. I could not reproduce that.

**Comment by fitdev** [comment]
I have not seen any crashes in VS 16.10 since 6244.

**Comment by wmjordan** [comment]
Thanks for the feedback.
I will release the new version then.

**Comment by wmjordan** [comment]
> > > 1. When the 2nd popup appear, it better to default the selection to the 1st item instead of nothing selected
> > >    Don't care about this. You can always press Enter key to activate the first item in the list if the caret is in the text box.
> 
> True, for the 1st item. However, for the 2nd item, I need to press down twice before pressing enter.

@fantasyz
There's a new beta which has implemented this. Now you can save a keystroke.
